pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```
@@ -0,0 +1,195 @@
+	membersCache   *membersCache
+	operatorClient v1helpers.OperatorClient
+	etcdClient     etcdcli.EtcdClient
+}
+
+func NewEtcdMemberListerController(operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &EtcdMemberListerController{
+		operatorClient: operatorClient,
+		etcdClient:     etcdClient,
+		membersCache:   newMembersCache(),
+	}
+	return factory.New().ResyncEvery(30*time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller"))
+}
+
+func (c *EtcdMemberListerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncEtcdMembers(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdMemberListerControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorSyncingEtcdMembers",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "EtcdMemberListerControllerDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "EtcdMembersSynced",
+	}))
+	if updateErr != nil {
+		syncCtx.Recorder().Warning("SyncEtcdMembersErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *EtcdMemberListerController) syncEtcdMembers(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+
+	var syncErrs []error
+	// build map from current live members
+	currentMembersMap, errs := c.currentMembersMap(members)
+	if len(errs) > 0 {
+		syncErrs = append(syncErrs, errs...)
```

Why are we continuing to check the cache here if we have errors from building the current members map?
Shouldn't we return the errors?

---

well I did not know what to do tbh :) 

So I can proceed with building the map while accumulating errors as I do now, or fail-fast once i encounter an error 

wdyt ? 

cc @tjungblu

---

fail-fast, but only degrade the operator after several consecutive failures.

---

done